### PR TITLE
Implement trusted channels

### PR DESCRIPTION
**Pull request by Infinisil**
Allows channels to declare a list of trusted channels with

    trustedChannels = [ "someChannel" ];

in their toplevel default.nix file

This has the effect that if a conflict were to exist between such a
channel and nixpkgs (and no others), the channel can be chosed without a
manual conflict resolution.

This should only be used for channels that are actually trusted, because
these can replace more nixpkgs dependencies without confirmation with
updates.

Docs still need to be updated

- [x] I have created a test to cover the new behavior.
- [ ] I have written and updated relevant documentation, including updating this
      pull request template if necessary. Note that we try to follow the
      [Divio documentation](https://documentation.divio.com/) of documentation.

### CODE REVIEW

**Comment by bpiv400** [comment]
Task linked: [CU-1cypecc Review: Implement trusted channels on flox-lib](https://app.clickup.com/t/1cypecc)

**Comment by limeytexan** [comment]
Thanks Silvan - happy to approve once we address the failing nixfmt test?

**Comment by Infinisil** [comment]
Yup, I'll fix this PR up soon so we can merge it :)

**Comment by Infinisil** [comment]
Although, thinking about this a bit more, I believe there's a better way to do this.

Currently if you depend on another channel, and that channel overrides a nixpkgs package, your own channel needs to specify a conflict resolution to choose between the nixpkgs version and the channel's version. But we can and should resolve this automatically _if_ the package to be resolved is only used by that very channel. I'll investigate if that's possible to implement.

**Comment by limeytexan** [comment]
> Although, thinking about this a bit more, I believe there's a better way to do this.

LGTM, but if you're considering a different/better way then do you still want to proceed with this? Happy either way - your call.

**Comment by Infinisil** [comment]
> Although, thinking about this a bit more, I believe there's a better way to do this.
> 
> Currently if you depend on another channel, and that channel overrides a nixpkgs package, your own channel needs to specify a conflict resolution to choose between the nixpkgs version and the channel's version. But we can and should resolve this automatically _if_ the package to be resolved is only used by that very channel. I'll investigate if that's possible to implement.

This turns out to be possible, though it's quite tricky, and not that useful. Here's an illustration of how it can work:
- Assume we have a channel A which defines packages X and Y, where Y depends on X, and X overrides the same-named nixpkgs package
- Now let's have channel B depending on channel A, making use of package Y, therefore including X transitively
- Currently you'd get a conflict resolution error for X, because we can't let a third-party channel override a nixpkgs package without confirmation.
- We can get around this by passing a different scope to channels A and B:
  - In A's scope, X is accessible without conflict resolution
  - If X were accessed in B's scope, X needs to be conflict resolved
- Avoiding this conflict can however notably only be done if X is only defined by a single channel

This isn't quite the full story though, but something that should be implemented along with this is the restriction that you can only depend on packages you _directly_ depend on. Because currently, if channel A depends on B depends on C, A can use the packages from C, even if it never declared a dependency on C. This ties into the above idea, because once we restrict channels to only be able to access packages from channels they directly depend on, we can have more guarantees about *waves hands*

_However_, this is actually quite useless, because it would only apply for packages that override one from nixpkgs, and that are only used from the channel that overrides them. But the fact that the package exists in nixpkgs also indicates that it's likely to be used by other channels.

So in short: I don't think this idea is worthwhile, too complex with too little benefit.

Instead, I'm taking another good look at this PR, which now seems like a much better solution.

**Comment by Infinisil** [comment]
Something that should be done before merging this PR is to introduce the notion of "safe" vs "unsafe" conflict resolutions: Imagine that we have channel A overriding nixpkgs package X. If we have another channel B depending on A, we currently get a conflict resolution like
```
error: The package pkgs.X is being used. However there are multiple channels which provide that package. No conflict resolution is currently provided. Valid options are \[ "A", "nixpkgs" \]. Set the conflict resolution for this package in .*/B/default.nix by copying one of the following lines to it:
conflictResolution.pkgs.X = "A";
conflictResolution.pkgs.X = "nixpkgs";
```

This error _sounds_ like B can freely choose either of them. However that is not the case exactly, because if `"nixpkgs"` is chosen, the override of X in A is completely ignored, even for packages in A itself, meaning that A's packages may break since they get a version of X that might not have the necessary properties. In comparison, if `"A"` is chosen, there aren't any problems, because the implicit assumption is that A's X contains some backwards compatible improvements over the nixpkgs version (this should be documented). We could present this to the user like this:

```
error: The package pkgs.X is being used. However there are multiple channels which provide that package. No conflict resolution is currently provided. Valid options are \[ "A", "nixpkgs" \]. Set the conflict resolution for this package in .*/B/default.nix by copying one of the following lines to it:
conflictResolution.pkgs.X = "A";
conflictResolution.pkgs.X = "nixpkgs"; # Unsafe: choosing this might break packages in A
```

We should even throw a warning if `"nixpkgs"` is chosen regardless, because ideally we don't want any unsafe overrides. There is an argument to be made that perhaps we don't want to entrust channel A with control over package X. However that inevitably leads to problems, because floxpkgs is essentially a distributed nixpkgs, which in order to have a consistent dependency set needs to give authority over attribute names to specific channels. There's no sensible way to resolve multiple channels giving conflicting definitions of a package. Throwing this warning weakly enforces this, because it encourages floxpkgs authors to communicate and compromise on the attribute names and overrides they use.

In fact, "package conflict resolution" should maybe be renamed to "package root authority"

---

Now this is important for this PR because we don't want to select unsafe resolutions, but we also want a trusted channel to be resolved in as many situations as possible. For example, right now this PR only allows a trusted channel to take over if it exactly only overrides nixpkgs. But it could also be that the trusted channel overrides a third channel, which itself overrides nixpkgs. The PR currently doesn't allow this, but it should. There may also be other cases that aren't covered, since this PR is rather naive at the moment.

I will try to implement this concept of safe vs unsafe overrides, and then base this PR off of that. This seems like the right solution to go for.

**Comment by Infinisil** [comment]
Oh but this leads me to another problem: If channel A defines package X overriding nixpkgs, and channel B depends on A and defines package X overriding A's X, then if A also depends on B, and A tries to build something, currently A's version of X is unconditionally used. This is a problem for B, because it won't be able to apply its extensions to X that it needs.

This tells me that the "if a channel defines a package, that package is the one used" methodology doesn't really work in general, because it can result in unsafe implicit conflict resolutions.

To fix this, we need to cleanly separate root conflict resolution and super conflict resolution. Up to now, these were doable in the single `conflictResolutions` definition, because super resolution was only necessary if a channel defined the package, at which point root resolution became pointless, as that channel's package was always the root.

What we need to do is to introduce a separate attribute, perhaps `superResolutions`, which specifies the super resolutions separately.

**Comment by Infinisil** [comment]
Will be superseded with #70 once that's done
